### PR TITLE
perf(core/executions): consider pipeline filters when fetching execut…

### PR DESCRIPTION
**Pull request by anotherchrisberry**
…ions

For applications with a very large list of pipeline executions, it's really cumbersome to view the history for a single pipeline beyond a couple of executions because Deck always fetches all the executions for the entire application. We already have an endpoint to retrieve executions by pipeline config ids, so we can use that to expedite retrieval.

The code is a little messy (we have to fetch the pipeline configs first, then correlate names to ids), but it's not bad, and ends up feeling a lot snappier. The only downside is there are more calls being made to the backend for smaller applications, too. But in my limited testing, this feels negligible on smaller apps.

There's also a fix in here for data source refresh. Right now, performing multiple `refresh(true)` calls on a data source will not necessarily result in the last one in being applied. This is especially noticeable on the executions as you toggle filters on and off. The fix is hacky because we're in a weird hybrid state between promises and observables in the data sources; nonetheless, it's functional and will be rewritten someday with `switchMap` and the world will be a better place.

@ajordens FYI
@jrsquared PTAL

### CODE REVIEW

File: app/scripts/modules/core/src/pipeline/manualExecution/manualPipelineExecution.html [hunk 1]
```diff
@@ -102,7 +102,7 @@ <h3 ng-if="!vm.command.pipeline">Select Pipeline</h3>
         <div class="form-group" ng-if="vm.hasRequiredParameters">
           <div class="col-md-4 col-md-offset-4">
             <em>* Required</em>
-
+          </div>
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/pipeline/manualExecution/manualPipelineExecution.html]: whoopsie (unrelated fix from last week)

File: app/scripts/modules/core/src/pipeline/service/execution.service.ts [hunk 2]
```diff
@@ -43,8 +47,23 @@ export class ExecutionService {
...
-    public getExecutions(applicationName: string): IPromise<IExecution[]> {
-      return this.getFilteredExecutions(applicationName);
+  /**
+   * Returns a filtered list of executions for the given application
+   * @param {string} applicationName the name of the application
+   * @param {Application} application: if supplied, and pipeline parameters are present on the filter model, the
+   * application will be used to correlate and filter the retrieved executions to only include those pipelines
+   * @return {<IExecution[]>}
+   */
+    public getExecutions(applicationName: string, application: Application = null): IPromise<IExecution[]> {
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/pipeline/service/execution.service.ts]: I don't love that the presence of the application argument triggers slightly different behavior (additional filtering), but not sure there's a clean way to do it without refactoring a bunch of other stuff.

File: app/scripts/modules/core/src/pipeline/service/execution.service.ts [hunk 3]
```diff
@@ -343,7 +374,7 @@ export class ExecutionService {
           application.executions.data.push(re);
         }
       });
-      if (updated) {
+      if (updated && !application.executions.reloadingForFilters) {
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/pipeline/service/execution.service.ts]: (this is the `mergeRunningExecutionsIntoExecutions` method)

If we are in the middle of reloading the filters, the `dataUpdated` call will clear the `reloadingForFilters` flag, which we don't want, as it appears the data is fully refreshed and, if the filters were just cleared, it'll show "no executions found for this pipeline" messages for any newly added pipelines.

File: app/scripts/modules/core/src/application/service/applicationDataSource.ts [hunk 4]
```diff
@@ -394,15 +407,17 @@ export class ApplicationDataSource implements IDataSourceConfig {
    * to true, and the promise will resolve immediately.
    *
    * If the data source is lazy and its "active" flag is set to false, any existing data will be cleared, the "loaded"
-   * flag will be set to false, adn the promise will resolve immediately.
+   * flag will be set to false, and the promise will resolve immediately.
```
Comment (jrsquared) [inline app/scripts/modules/core/src/application/service/applicationDataSource.ts]: best change

File: app/scripts/modules/core/src/application/service/applicationDataSource.ts [hunk 5]
```diff
@@ -394,15 +407,17 @@ export class ApplicationDataSource implements IDataSourceConfig {
...
    * If the data source is in the process of loading, the promise will resolve immediately. This behavior can be
    * overridden by calling "refresh(true)".
    *
    * @param forceRefresh
    * @returns {any}
    */
-  public refresh(forceRefresh?: boolean): IPromise<any> {
+  public refresh(forceRefresh?: boolean): IPromise<void> {
+    const deferred = this.$q.defer<void>();
+    this.refreshQueue.push(deferred);
```
Comment (jrsquared) [inline app/scripts/modules/core/src/application/service/applicationDataSource.ts]: clever

**Comment by jrsquared** [comment]
LGTM
